### PR TITLE
Update MySQL to Version 8.0

### PR DESCRIPTION
**Pull request by Kitzunu**
## CHANGES PROPOSED:
-  Deprecate MySQL 5.6
-  Allow usage of MySQL 8.0

## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/2895

## TESTS PERFORMED:
- Builds
- Tested in-game, so far no issues

## HOW TO TEST THE CHANGES:

1. build
2. 🤷 

## KNOWN ISSUES:

- [ ] DB Assembler is broken due to this. (Old query that isn't allowed anymore, I don't know how to fix it. Check linked issue for ref)

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/cmake/macros/FindMySQL.cmake [hunk 1]
```diff
@@ -143,14 +143,14 @@ find_path(MYSQL_INCLUDE_DIR
     /usr/local/include
     /usr/local/include/mysql
     /usr/local/mysql/include
+    "C:/Program Files/MySQL/MySQL Server 8.0/include"
     "C:/Program Files/MySQL/MySQL Server 5.7/include"
-    "C:/Program Files/MySQL/MySQL Server 5.6/include"
```
Comment (FrancescoBorzi) [inline src/cmake/macros/FindMySQL.cmake:147]: is deprecating MySQL 5.6 needed in order to support 8.0? if yes, ok for me. Otherwise we can keep it until February 2021

File: src/server/game/Groups/GroupMgr.cpp [hunk 2]
```diff
@@ -91,19 +91,19 @@ void GroupMgr::LoadGroups()
         uint32 oldMSTime = getMSTime();
 
         // Delete all groups whose leader does not exist
-        CharacterDatabase.DirectExecute("DELETE FROM groups WHERE leaderGuid NOT IN (SELECT guid FROM characters)");
+        CharacterDatabase.DirectExecute("DELETE FROM `groups` WHERE leaderGuid NOT IN (SELECT guid FROM characters)");
         // Delete all groups with less than 2 members (or less than 1 for lfg groups)
-        CharacterDatabase.DirectExecute("DELETE groups FROM groups LEFT JOIN ((SELECT guid, count(*) as cnt FROM group_member GROUP BY guid) t) ON groups.guid = t.guid WHERE t.guid IS NULL OR (t.cnt<=1 AND groups.groupType <> 12)");
+        CharacterDatabase.DirectExecute("DELETE FROM `groups` WHERE guid NOT IN (SELECT guid FROM group_member GROUP BY guid HAVING COUNT(guid) > 1)");
```
Comment (FrancescoBorzi) [inline src/server/game/Groups/GroupMgr.cpp:96]: @Kitzunu  can you explain why this change?

File: src/server/game/Groups/GroupMgr.cpp [hunk 3]
```diff
@@ -91,19 +91,19 @@ void GroupMgr::LoadGroups()
...
         // Delete invalid lfg_data
-        CharacterDatabase.DirectExecute("DELETE lfg_data FROM lfg_data LEFT JOIN groups ON lfg_data.guid = groups.guid WHERE groups.guid IS NULL OR groups.groupType <> 12");
+        CharacterDatabase.DirectExecute("DELETE FROM lfg_data WHERE guid IN (SELECT guid FROM `groups` GROUP BY guid HAVING COUNT(guid) <> 12)");
```
Comment (FrancescoBorzi) [inline src/server/game/Groups/GroupMgr.cpp:98]: same as above

File: src/server/game/Groups/GroupMgr.cpp (hunk 2 above)
Comment (Kitzunu) [reply src/server/game/Groups/GroupMgr.cpp:96]: I change because otherwise we got worldserver startup errors.
I have not actually tested that change myself

File: src/server/game/Groups/GroupMgr.cpp (hunk 3 above)
Comment (Kitzunu) [reply src/server/game/Groups/GroupMgr.cpp:98]: I change because otherwise we got worldserver startup errors.
I have not actually tested that change myself

File: src/cmake/macros/FindMySQL.cmake (hunk 1 above)
Comment (Kitzunu) [reply src/cmake/macros/FindMySQL.cmake:147]: No, but if I understand what I have read correctly, these changes can't be supported by 5.6.

But if someone have 5.6 and can test the changes without deprecating it, feel free. I cannot test that myself since I fully upgraded to 8.0

File: src/server/game/Groups/GroupMgr.cpp (hunk 3 above)
Comment (Kitzunu) [reply src/server/game/Groups/GroupMgr.cpp:98]: Without this change, we get this error in worldserver which prevents it from starting
```
2020-04-16 20:50:54 SQL: DELETE lfg_data FROM lfg_data LEFT JOIN groups ON lfg_data.guid = groups.guid WHERE groups.guid IS NULL OR groups.groupType <> 12
2020-04-16 20:50:54 ERROR: [1064] You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'groups ON lfg_data.guid = groups.guid WHERE groups.guid IS NULL OR groups.groupT' at line 1
```

File: src/server/game/Groups/GroupMgr.cpp (hunk 2 above)
Comment (Kitzunu) [reply src/server/game/Groups/GroupMgr.cpp:96]: Without this change, we get this error in worldserver which prevents it from starting:
```
2020-04-16 20:49:34 SQL: DELETE groups FROM groups LEFT JOIN ((SELECT guid, count(*) as cnt FROM group_member GROUP BY guid) t) ON groups.guid = t.guid WHERE t.guid IS NULL OR (t.cnt<=1 AND groups.groupType <> 12)
2020-04-16 20:49:34 ERROR: [1064] You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'groups FROM groups LEFT JOIN ((SELECT guid, count(*) as cnt FROM group_member GR' at line 1
```

File: src/cmake/macros/FindMySQL.cmake (hunk 1 above)
Comment (Viste) [reply src/cmake/macros/FindMySQL.cmake:147]: 8.0 breaks 5.6 its true

Comment (Kitzunu) [reply src/cmake/macros/FindMySQL.cmake:147]: Yea that's what I thought too after what I read online, but wasn't too sure!
Thank you for confirming!!

**Comment by Viste** [comment]
be aware it also can break mariadb(dont sure)
can test it but dont know when

**Comment by Kitzunu** [comment]
It would be awesome if it could be checked.
Perhaps I have missed something here to remove from 5.6 or to add for 8.0 MySQL too.

All I focused on here was to get the build (windows) working and get all worldserver.exe errors fixed

**Comment by Viste** [comment]
with mariadb 10.4 same error as in travis :C
```
[ 15%] Building CXX object src/common/CMakeFiles/common.dir/Database/Transaction.cpp.o
/home/viste/dev-wow/azerothcore-wotlk/src/common/Database/QueryResult.cpp:69:30: fatal error: assigning to 'my_bool *' (aka 'char *') from incompatible type 'bool *'
        m_rBind[i].is_null = &m_isNull[i];
                             ^~~~~~~~~~~~
1 error generated.
make[2]: *** [src/common/CMakeFiles/common.dir/build.make:490: src/common/CMakeFiles/common.dir/Database/QueryResult.cpp.o] Ошибка 1
make[2]: *** Ожидание завершения заданий…
make[1]: *** [CMakeFiles/Makefile2:977: src/common/CMakeFiles/common.dir/all] Ошибка 2
make: *** [Makefile:150: all] Ошибка 2
```

File: src/server/game/Groups/GroupMgr.cpp (hunk 3 above)
Comment (BarbzYHOOL) [reply src/server/game/Groups/GroupMgr.cpp:98]: that's why we shouldn't start big PRs like that if not sure

**Comment by Viste** [comment]
i think we should add -DWITHMYSQL8 option to cmake and ifdef changes occurred to this in core
but don't know what to do with sql

**Comment by Kitzunu** [comment]
I will leave this PR, if anyone more experienced can fix any errors between Maria and MySQL feel free to take over.

I am happy and proud that I actually managed to make it work locally, that was a big step for me :)

File: src/cmake/macros/FindMySQL.cmake (hunk 1 above)
Comment (FrancescoBorzi) [reply src/cmake/macros/FindMySQL.cmake:147]: ok, then it's fine for me. Supporting 8.0 it's better than 5.6. Also @Viste let us know if it's ok for Maria too, because that's even more important

File: src/server/game/Groups/GroupMgr.cpp (hunk 2 above)
Comment (FrancescoBorzi) [reply src/server/game/Groups/GroupMgr.cpp:96]: @Kitzunu yes but what about the logic change here? I don't see this anymore for example: `groups.groupType <> 12`

File: src/server/game/Groups/GroupMgr.cpp (hunk 3 above)
Comment (FrancescoBorzi) [reply src/server/game/Groups/GroupMgr.cpp:98]: @Kitzunu I understand that you changed in order to avoid SQL errors, but I don't understand how the logic is changed here

**Comment by FrancescoBorzi** [comment]
It would be nice to understand what parts are breaking MariaDB

**Comment by Viste** [comment]
bool shit :DDD so funny

File: src/common/Database/QueryResult.h [hunk 4]
```diff
@@ -84,7 +84,7 @@ class PreparedResultSet
         MYSQL_STMT* m_stmt;
         MYSQL_RES* m_res;
 
-        my_bool* m_isNull;
+        bool* m_isNull;
```
Comment (FrancescoBorzi) [inline src/common/Database/QueryResult.h:87]: @Kitzunu try to revert this change + apply this PR:

https://github.com/azerothcore/azerothcore-wotlk/pull/3060

File: src/cmake/macros/FindMySQL.cmake [hunk 5]
```diff
@@ -181,18 +181,18 @@ if( WIN32 )
...
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\MySQL AB\\MySQL Server 5.7;Location]/lib"
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\MySQL AB\\MySQL Server 5.7;Location]/lib/opt"
-      "[HKEY_LOCAL_MACHINE\\SOFTWARE\\MySQL AB\\MySQL Server 5.6;Location]/lib"
-      "[HKEY_LOCAL_MACHINE\\SOFTWARE\\MySQL AB\\MySQL Server 5.6;Location]/lib/opt"
+      "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\MySQL AB\\MySQL Server 8.0;Location]/lib"
+      "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\MySQL AB\\MySQL Server 8.0;Location]/lib/opt"
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\MySQL AB\\MySQL Server 5.7;Location]/lib"
       "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\MySQL AB\\MySQL Server 5.7;Location]/lib/opt"
-      "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\MySQL AB\\MySQL Server 5.6;Location]/lib"
-      "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\MySQL AB\\MySQL Server 5.6;Location]/lib/opt"
```
Comment (FrancescoBorzi) [inline src/cmake/macros/FindMySQL.cmake:195]: @Kitzunu do we have this part on master? do we need it? I don't have Windows so I can't tell

Comment (Kitzunu) [reply src/cmake/macros/FindMySQL.cmake:195]: I'll add the 8.0 part and keep the 5.6 part :)
